<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN" "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
	<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
	<link rel="stylesheet" href="configuration/report.css" type="text/css" />
	<title>RnBeads report</title>
	<script src="configuration/report.js" type="text/javascript"></script>
</head>

<body style="background:url(configuration/RnBeads.png) no-repeat 5px 25px;">

<h1>Differential Methylation</h1>


<h2 class="collapsed" id="hsection1" onclick="toggleSection('1')">Parameter Overview</h2>
<div id="section1" style="display:none">
<p>The table below lists the options of the executed module.</p>
<table class="tindex">
<thead>
<tr>
	<th>Option</th>
	<th>Value</th>
</tr>
</thead>
<tbody>
<tr>
	<td>analyze.sites</td>
	<td>yes</td>
</tr>
<tr class="darker">
	<td>differential.report.sites</td>
	<td>yes</td>
</tr>
<tr>
	<td>region.types</td>
	<td><span class="missing">default</span></td>
</tr>
<tr class="darker">
	<td>differential.permutations</td>
	<td>0</td>
</tr>
<tr>
	<td>differential.comparison.columns</td>
	<td>diseaseState</td>
</tr>
<tr class="darker">
	<td>differential.comparison.columns.all.pairwise</td>
	<td><span class="missing">default</span></td>
</tr>
<tr>
	<td>columns.pairing</td>
	<td><span class="missing">default</span></td>
</tr>
<tr class="darker">
	<td>differential.site.test.method</td>
	<td>limma</td>
</tr>
<tr>
	<td>covariate.adjustment.columns</td>
	<td>age, gender, smokingStatus</td>
</tr>
<tr class="darker">
	<td>differential.adjustment.sva</td>
	<td>yes</td>
</tr>
<tr>
	<td>differential.adjustment.celltype</td>
	<td>yes</td>
</tr>
<tr class="darker">
	<td>differential.enrichment</td>
	<td>no</td>
</tr>
</tbody>
</table>
</div>

<h2 class="expanded" id="hsection2" onclick="toggleSection('2')">Introduction: Differential Methylation of Sample Groups</h2>
<div id="section2" style="display:block">
<p>Differential methylation analysis was conducted on site and region level according to the sample groups specified in the analysis.</p>

<h3 class="expanded" id="hsection3" onclick="toggleSection('3')">Comparisons</h3>
<div id="section3" style="display:block">
<p>The following comparisons were made:</p>
<ul>
	<li>Normal vs. RheumatoidArthritis (based on diseaseState) [38 vs. 12]</li>
</ul>
<p>The table below summarizes information on the comparisons.</p>
<div class="restricted"><table class="tabdata">
<tr>
	<td class="header"></td>
	<td class="header">comparison</td>
	<td class="header">adjustment</td>
	<td class="header">covariateTable</td>
</tr>
<tr>
	<td class="header">1</td>
	<td class="header">Normal vs. RheumatoidArthritis (based on diseaseState)</td>
	<td class="centered">age,gender,smokingStatus</td>
	<td class="centered"><a href="differential_methylation_data/adjTable_1.csv">csv</a></td>
</tr>
</table></div>

</div>

<h3 class="expanded" id="hsection4" onclick="toggleSection('4')">P-values</h3>
<div id="section4" style="display:block">
<p>In the following anlyses, p-values on the site level were computed using the <code>limma</code> method. I.e. hierarchical linear models from the <a href=http://bioinf.wehi.edu.au/limma/>limma</a> package were employed and fitted using an empirical Bayes approach on derived M-values.</p>
</div>
</div>

<h2 class="expanded" id="hsection5" onclick="toggleSection('5')">Site Level</h2>
<div id="section5" style="display:block">
<p>Differential methylation on the site level was computed based on a variety of metrics. 
						  Of particular interest for the following plots and analyses are the following quantities for each site:
						  a) the difference in mean methylation levels of the two groups being compared, b) the quotient in mean methylation and
						  c) a statistical test (limma or t-test depending on the settings) assessing whether the methylation values in the two groups originate from distinct distributions.
						  Additionally each site was assigned a rank based on each of these three criteria. A combined rank is computed as the maximum (i.e. worst)
						  rank among the three ranks. The smaller the combined rank for a site, the more evidence for differential methylation it exhibits.
						  This section includes scatterplots of the site group means as well as volcano plots
						  of each pairwise comparison colored according to the combined ranks or p-values of a given site.</p>
<p>The following rank cutfoffs have been automatically selected for the analysis of differentially methylated sites:</p>
<div class="restricted"><table class="tabdata">
<tr>
	<td class="header"></td>
	<td class="header">Rank Cutoff</td>
</tr>
<tr>
	<td class="header">Normal vs. RheumatoidArthritis (based on diseaseState)</td>
	<td class="centered">114314</td>
</tr>
</table></div>

<div class="figure">
	<table class="tabsettings">
	<tr>
		<td>comparison</td>
		<td>
			<select disabled="disabled" id="fig1setting3" onchange="updateFigure('fig1')">
				<option selected="selected" value="cmp1">Normal vs. RheumatoidArthritis (based on diseaseState)</option>
			</select>
		</td>
	</tr>
	<tr>
		<td>differential methylation measure</td>
		<td>
			<select id="fig1setting4" onchange="updateFigure('fig1')">
				<option selected="selected" value="fdrAdjPval">FDR adjusted p-value &lt; 0.05</option>
				<option value="rc1">combined rank among the 100 best ranking sites</option>
				<option value="rc2">combined rank among the 1000 best ranking sites</option>
				<option value="rc3">combined rank among the 10000 best ranking sites</option>
				<option value="rc4">combined rank among the 1e+05 best ranking sites</option>
				<option value="rcAuto">automatically selected rank cutoff</option>
				<option value="rankGradient">combined rank based gradient</option>
			</select>
		</td>
	</tr>
	</table>
	<br />
	<a href="differential_methylation_images/diffMeth_site_cmp1_fdrAdjPval_high_resolution.png" id="fig1imagehigh"><img alt="Figure 1" id="fig1image" height="700" src="differential_methylation_images/diffMeth_site_cmp1_fdrAdjPval.png" width="700" /></a>
	<p class="caption">
		<a id="fig1caption" onmousedown="showHide('fig1description')" onmouseout="doCursor()" onmouseover="doCursor()">Figure 1</a>
	</p>
	<p class="figdescription" id="fig1description">Scatterplot for differential methylation (sites). If the selected criterion is not <code>rankGradient</code>:
		The transparency corresponds to point density. If the number of points exceeds 2e+06 then the number of points for density estimation is reduced to that number by random sampling.The1% of the points in the sparsest populated plot regions are drawn explicitly (up to a maximum of 10000 points).Additionally, the colored points represent differentially methylated sites (according to the selected criterion). 
		If the selected criterion is <code>rankGradient</code>: median combined ranks accross hexagonal bins are shown
		as a gradient according to the color legend.</p>
</div>

<div class="figure">
	<table class="tabsettings">
	<tr>
		<td>comparison</td>
		<td>
			<select disabled="disabled" id="fig2setting4" onchange="updateFigure('fig2')">
				<option selected="selected" value="cmp1">Normal vs. RheumatoidArthritis (based on diseaseState)</option>
			</select>
		</td>
	</tr>
	<tr>
		<td>difference metric</td>
		<td>
			<select id="fig2setting5" onchange="updateFigure('fig2')">
				<option selected="selected" value="diff">Difference</option>
				<option value="quot">Quotient</option>
			</select>
		</td>
	</tr>
	<tr>
		<td>significance metric</td>
		<td>
			<select id="fig2setting6" onchange="updateFigure('fig2')">
				<option selected="selected" value="pVal">p-value</option>
				<option value="pValAdj">adjusted p-value</option>
				<option value="quotSig">Quotient (only meaningful if 'Difference' is selected above)</option>
			</select>
		</td>
	</tr>
	</table>
	<br />
	<a href="differential_methylation_images/diffMeth_site_volcano_cmp1_diff_pVal_high_resolution.png" id="fig2imagehigh"><img alt="Figure 2" id="fig2image" height="700" src="differential_methylation_images/diffMeth_site_volcano_cmp1_diff_pVal.png" width="700" /></a>
	<p class="caption">
		<a id="fig2caption" onmousedown="showHide('fig2description')" onmouseout="doCursor()" onmouseover="doCursor()">Figure 2</a>
	</p>
	<p class="figdescription" id="fig2description">Volcano plot for differential methylation quantified by various metrics. Color scale according to
					combined ranking.</p>
</div>


<h3 class="expanded" id="hsection6" onclick="toggleSection('6')">Differential Methylation Tables</h3>
<div id="section6" style="display:block">
<p>A tabular overview of measures for differential methylation on the site level for the individual comparisons are provided in this section.
					  Below, a brief explanation of the different columns can be found:</p>
<ul>
	<li>id: site id</li>
	<li>Chromosome: chromosome of the site</li>
	<li>Start: start coordinate of the site</li>
	<li>Strand: strand of the site</li>
	<li>mean.g1,mean.g2: (where g1 and g2 is replaced by the respective group names in the table) mean methylation in each of the two groups</li>
	<li>mean.diff: difference in methylation means between the two groups: mean.g1-mean.g2. In case of paired analysis, it is the mean of the pairwise differences.</li>
	<li>mean.quot.log2: log2 of the quotient in methylation: log2((mean.g1+epsilon)/(mean.g2+epsilon)), where epsilon:=0.01. In case of paired analysis, it is the mean of the pairwise quotients.</li>
	<li>diffmeth.p.val: p-value obtained from linear models employed in the limma package (or alternatively from a two-sided Welch t-test; which type of p-value is computed is specified in the differential.site.test.method option).</li>
	<li>max.g1,max.g2: maximum methylation level in group 1 and 2 respectively</li>
	<li>min.g1,min.g2: minimum methylation level in group 1 and 2 respectively</li>
	<li>sd.g1,sd.g2: standard deviation of methylation levels</li>
	<li>min.diff: Minimum of 0 and the smallest pairwise difference between samples of the two groups</li>
	<li>diffmeth.p.adj.fdr: FDR adjusted p-value of all sites</li>
	<li>combinedRank: mean.diff, mean.quot.log2 and diffmeth.p.val are ranked for all sites. This aggregates them using the maximum, i.e. worst rank of a site among the three measures</li>
	<li>num.na.g1,num.na.g2: number of NA methylation values for groups 1 and 2 respectively</li>
	<li>mean.covg.g1,mean.covg.g2: mean coverage of groups 1 and 2 respectively (In case of Infinium array methylation data, coverage is defined as combined beadcount.)</li>
	<li>min.covg.g1,min.covg.g2: minimum coverage of groups 1 and 2 respectively</li>
	<li>max.covg.g1,max.covg.g2: maximum coverage of groups 1 and 2 respectively</li>
	<li>covg.thresh.nsamples.g1,covg.thresh.nsamples.g2: number of samples in group 1 and 2 respectively exceeding the coverage threshold (5) for this site.</li>
</ul>
<p>The tables for the individual comparisons can be found here:
<ul>
<li><a href="differential_methylation_data/diffMethTable_site_cmp1.csv">Normal vs. RheumatoidArthritis (based on diseaseState)</a></li>
</ul></p>
</div>
</div>

<h2 class="expanded" id="hsection7" onclick="toggleSection('7')">Region Level</h2>
<div id="section7" style="display:block">
<p>Differential methylation on the region level was computed based on a variety of metrics. Of particular interest for the following plots and analyses are the following quantities for each region: the mean difference in means across all sites in a region of the two groups being compared and the mean of quotients in mean methylation as well as a combined p-value calculated from all site p-values in the region [<a href="#ref1">1</a>].  Additionally each region was assigned a rank based on each of these three criteria. A combined rank is computed as the maximum (i.e. worst) value among the three ranks. The smaller the combined rank for a region, the more evidence for differential methylation it exhibits. Regions were defined based on the region types specified in the analysis. This section includes scatterplots of the region group means as well as volcano plots of each pairwise comparison colored according to the combined rank of a given region.</p>
<p>The following rank cutfoffs have been automatically selected for the analysis of differentially methylated regions:</p>
<div class="restricted"><table class="tabdata">
<tr>
	<td class="header"></td>
	<td class="header">tiling</td>
	<td class="header">genes</td>
	<td class="header">promoters</td>
	<td class="header">cpgislands</td>
</tr>
<tr>
	<td class="header">Normal vs. RheumatoidArthritis (based on diseaseState)</td>
	<td class="centered">7120</td>
	<td class="centered">455</td>
	<td class="centered">1532</td>
	<td class="centered">87</td>
</tr>
</table></div>

<div class="figure">
	<table class="tabsettings">
	<tr>
		<td>comparison</td>
		<td>
			<select disabled="disabled" id="fig3setting3" onchange="updateFigure('fig3')">
				<option selected="selected" value="cmp1">Normal vs. RheumatoidArthritis (based on diseaseState)</option>
			</select>
		</td>
	</tr>
	<tr>
		<td>regions</td>
		<td>
			<select id="fig3setting4" onchange="updateFigure('fig3')">
				<option selected="selected" value="tiling">tiling</option>
				<option value="genes">genes</option>
				<option value="promoters">promoters</option>
				<option value="cpgislands">cpgislands</option>
			</select>
		</td>
	</tr>
	<tr>
		<td>differential methylation measure</td>
		<td>
			<select id="fig3setting5" onchange="updateFigure('fig3')">
				<option selected="selected" value="fdrAdjPval">FDR adjusted p-value < 0.05</option>
				<option value="rc1">combined rank among the 100 best ranking regions</option>
				<option value="rc2">combined rank among the 500 best ranking regions</option>
				<option value="rc3">combined rank among the 1000 best ranking regions</option>
				<option value="rcAuto">automatically selected rank cutoff</option>
				<option value="rankGradient">combined rank based gradient</option>
			</select>
		</td>
	</tr>
	</table>
	<br />
	<a href="differential_methylation_images/diffMeth_region_cmp1_tiling_fdrAdjPval_high_resolution.png" id="fig3imagehigh"><img alt="Figure 3" id="fig3image" height="700" src="differential_methylation_images/diffMeth_region_cmp1_tiling_fdrAdjPval.png" width="700" /></a>
	<p class="caption">
		<a id="fig3caption" onmousedown="showHide('fig3description')" onmouseout="doCursor()" onmouseover="doCursor()">Figure 3</a>
	</p>
	<p class="figdescription" id="fig3description">Scatterplot for differential methylation (regions). If the selected criterion is not <code>rankGradient</code>:
		The transparency corresponds to point density. The 1% of the points in the sparsest populated plot regions are drawn explicitly.
		Additionally, the colored points represent differentially methylated regions (according to the selected criterion). 
		If the selected criterion is <code>rankGradient</code>: median combined ranks accross hexagonal bins are shown
		as a gradient according to the color legend.</p>
</div>

<div class="figure">
	<table class="tabsettings">
	<tr>
		<td>comparison</td>
		<td>
			<select disabled="disabled" id="fig4setting4" onchange="updateFigure('fig4')">
				<option selected="selected" value="cmp1">Normal vs. RheumatoidArthritis (based on diseaseState)</option>
			</select>
		</td>
	</tr>
	<tr>
		<td>regions</td>
		<td>
			<select id="fig4setting5" onchange="updateFigure('fig4')">
				<option selected="selected" value="tiling">tiling</option>
				<option value="genes">genes</option>
				<option value="promoters">promoters</option>
				<option value="cpgislands">cpgislands</option>
			</select>
		</td>
	</tr>
	<tr>
		<td>difference metric</td>
		<td>
			<select id="fig4setting6" onchange="updateFigure('fig4')">
				<option selected="selected" value="diff">Difference</option>
				<option value="quot">Quotient</option>
			</select>
		</td>
	</tr>
	<tr>
		<td>significance metric</td>
		<td>
			<select id="fig4setting7" onchange="updateFigure('fig4')">
				<option selected="selected" value="pVal">combined p-value</option>
				<option value="pValAdj">adjusted combined p-value</option>
				<option value="quotSig">Quotient (only meaningful if 'Difference' is selected above)</option>
			</select>
		</td>
	</tr>
	</table>
	<br />
	<a href="differential_methylation_images/diffMeth_region_volcano_cmp1_tiling_diff_pVal_high_resolution.png" id="fig4imagehigh"><img alt="Figure 4" id="fig4image" height="700" src="differential_methylation_images/diffMeth_region_volcano_cmp1_tiling_diff_pVal.png" width="700" /></a>
	<p class="caption">
		<a id="fig4caption" onmousedown="showHide('fig4description')" onmouseout="doCursor()" onmouseover="doCursor()">Figure 4</a>
	</p>
	<p class="figdescription" id="fig4description">Volcano plot for differential methylation quantified by various metrics. Color scale according to
		combined ranking.</p>
</div>


<h3 class="expanded" id="hsection8" onclick="toggleSection('8')">Differential Methylation Tables</h3>
<div id="section8" style="display:block">
<p>A tabular overview of measures for differential methylation on the region level for the individual comparisons are provided in this section.</p>
<ul>
	<li>id: region id</li>
	<li>Chromosome: chromosome of the region</li>
	<li>Start: start coordinate of the region</li>
	<li>End: end coordinate of the region</li>
	<li>[symbol]: associated gene symbol to the given region [only valid for gene associated regions]</li>
	<li>[entrezID]: Entrez ID of the gene associated with the region [only valid for gene associated regions]</li>
	<li>mean.mean.g1,mean.mean.g2: (where g1 and g2 is replaced by the respective group names in the table) mean of mean methylation levels for group 1 and 2 across all sites in a region</li>
	<li>mean.mean.diff: Mean difference in means across all sites in a region</li>
	<li>mean.mean.quot.log2: log2 of the mean quotient in means across all sites in a region</li>
	<li>comb.p.val: Combined p-value aggregating p-values of all sites in the region using a generalization of Fisher's method [<a href="#ref1">1</a>]</li>
	<li>comb.p.adj.fdr: FDR adjusted combined p-value</li>
	<li>combinedRank: mean.mean.diff, mean.mean.quot.log2 and comb.p.val are ranked for all regions. This column aggregates them using the maximum, i.e. worst rank of a site among the three measures</li>
	<li>num.sites: number of sites associated with the region</li>
	<li>mean.num.na.g1,mean.num.na.g2: Mean number of NA methylation values accross all sites in group 1 and group 2 respectively</li>
	<li>mean.mean.covg.g1,mean.mean.covg.g2: Mean value of mean coverage values (across all samples in a group) across all sites in a region</li>
	<li>mean.nsamples.covg.thresh.g1,mean.nsamples.covg.thresh.g2: mean number of samples (accross all considered sites) that have a coverage larger than 5 for the site in group 1 and group 2 respectively</li>
</ul>
<p>The tables for the individual comparisons can be found here:</p>
<div class="restricted"><table class="tabdata">
<tr>
	<td class="header"></td>
	<td class="header">tiling</td>
	<td class="header">genes</td>
	<td class="header">promoters</td>
	<td class="header">cpgislands</td>
</tr>
<tr>
	<td class="header">Normal vs. RheumatoidArthritis (based on diseaseState)</td>
	<td class="centered"><a href="differential_methylation_data/diffMethTable_region_cmp1_tiling.csv">csv</a></td>
	<td class="centered"><a href="differential_methylation_data/diffMethTable_region_cmp1_genes.csv">csv</a></td>
	<td class="centered"><a href="differential_methylation_data/diffMethTable_region_cmp1_promoters.csv">csv</a></td>
	<td class="centered"><a href="differential_methylation_data/diffMethTable_region_cmp1_cpgislands.csv">csv</a></td>
</tr>
</table></div>

</div>

<h3 class="expanded" id="hsection9" onclick="toggleSection('9')">Enrichment Analysis</h3>
<div id="section9" style="display:block">
<p>No Enrichment Analysis was conducted</p>
</div>
</div>

<h2 class="expanded" id="hsection10" onclick="toggleSection('10')">References</h2>
<div id="section10" style="display:block">
<ol>
	<li><a id="ref1">Makambi, K. (2003) Weighted inverse chi-square method for correlated significance tests. <i>Journal of Applied Statistics</i>, <b>30</b>(2), 225234</a></li>
</ol>
</div>

<div id="copyright">
	<div id="rnbeads">
	This report was generated on 2016-07-08 by <a href="http://rnbeads.mpi-inf.mpg.de/">RnBeads</a> version 1.5.0.
	</div>
	<div id="validlogo">
		<a href="http://validator.w3.org/check?uri=referer"><img alt="Valid XHTML 1.1" height="31" src="http://www.w3.org/Icons/valid-xhtml11-blue" width="88" /></a>
		<a href="http://jigsaw.w3.org/css-validator/check/referer"><img alt="Valid CSS" height="31" src="http://jigsaw.w3.org/css-validator/images/vcss-blue" width="88" /></a>
	</div>
</div>
</body>
</html>
